src/main/java/org/folio/auth/authtokenmodule/MainVerticle.java

```diff
@@ -456,316 +432,272 @@ private void handleAuthorize(RoutingContext ctx) {
-                        "__" + df.format(now))
-                .put("tenant", tenant)
-                .put("dummy", true);
-      } catch(Exception e) {
-        endText(ctx, 500,  "Error creating dummy token: ", e);
-        return;
-      }
-      try {
-        candidateToken = tokenCreator.createJWTToken(dummyPayload.encode());
-      } catch(Exception e) {
-        endText(ctx, 500, "Error creating candidate token: ", e);
-        return;
-      }
-    }
-
     final String authToken = candidateToken;
     logger.debug("Final authToken is {}", authToken);
-    final String errMsg = "Invalid token";
-    try {
-      if (!tokenCache.contains(authToken)) {
-        tokenCreator.checkJWTToken(authToken);
-        tokenCache.add(authToken);
-      }
-    } catch (ParseException p) {
-      logger.error("Malformed token: {}", authToken, p);
-      endText(ctx, 401, errMsg);
-      return;
-    } catch (JOSEException j) {
-      logger.error("Unable to verify token token {}, {}", authToken, j.getMessage());
-      endText(ctx, 401, errMsg);
-      return;
-    } catch (BadSignatureException b) {
-      logger.error("Unsupported JWT format", b);
-      endText(ctx, 401, errMsg);
-      return;
-    }
 
-    JsonObject tokenClaims = getClaims(authToken);
+    Future<Token> tokenValidationResult = Token.validate(authToken, tokenCreator, ctx.request());
 
-    String username = tokenClaims.getString("sub");
-    String jwtTenant = tokenClaims.getString("tenant");
-
-    if (jwtTenant == null || !jwtTenant.equals(tenant)) {
-      logger.error("Expected tenant: {}, got tenant {}", tenant, jwtTenant);
-      endText(ctx, 403, "Invalid token for access");
+    tokenValidationResult.onFailure(h -> {
+      var e = (TokenValidationException)h;
+      logger.error("Invalid token: {}", e.toString());
+      endText(ctx, e.httpResponseCode, "Invalid token");
       return;
-    }
-
-    String tokenUserId = tokenClaims.getString(TOKEN_USER_ID_FIELD);
-    if (tokenUserId != null) {
-      if (userId != null) {
-        if (!userId.equals(tokenUserId)) {
-          endText(ctx, 403,
-           "Payload user id of '" + tokenUserId + "' does not match expected value");
-          return;
+    });
+    
+    tokenValidationResult.onSuccess(h -> {
+      Token token = (Token)h;
+      logger.debug("Validated token of type: {}", token.getClaims().getString("type"));
+
+      String username = token.getClaims().getString("sub");
+
+      // TODO I believe this block solves the issue that Adam found with functional testing.
+      // My approach of just making userId be equal to the X-Okapi-User-Id didn't
+      // give what is here and in the existing code called the finalUserId a value
+      // in this specific case of the tokenUserId having a value _and_ the X-Okapi-User-Id
+      // being null. I'd likek to better understand the circumstances under which these
+      // two variables can have this state and what it means.
```

Comment:
Okapi uses that same logic: https://github.com/folio-org/okapi/blob/v4.9.0/okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java#L280